### ISSUE TITLE
User Story Testing with Novice QUIC Users

### ISSUE THREAD

**Issue by hlandau** [openssl/project#253]
Get test users without prior knowledge of QUIC or OpenSSL QUIC to try to use OpenSSL QUIC and produce a journal detailing frustration and confusion points they encountered in that process. Our new developers are likely to be suitable guinea pigs for this purpose.

**Comment by jamuir** [comment]
So far I have reviewed README-QUIC.md and part of the other docs it points to:

- https://www.openssl.org/docs/manmaster/man7/openssl-quic.html
- the four sections on QUIC from the OpenSSL Guide:
    - https://www.openssl.org/docs/manmaster/man7/ossl-guide-quic-introduction.html
    - https://www.openssl.org/docs/manmaster/man7/ossl-guide-quic-client-block.html
    - https://www.openssl.org/docs/manmaster/man7/ossl-guide-quic-multi-stream.html
    - https://www.openssl.org/docs/manmaster/man7/ossl-guide-quic-client-non-block.html

My comments are in the attached markdown file.
[jm-journal-part1.md](https://github.com/openssl/project/files/13110860/jm-journal-part1.md)

I have opened the following PR with some minor doc and demo-code fixes:

https://github.com/openssl/openssl/pull/22483

Generally, I think the QUIC docs and demos are very good.  Stylistically, my preference would be to use more commas in the prose :-)  I find commas make for easier reading.

Two things to point out (both of which are mentioned in the attached file): 

1. when using `openssl s_client -quic ...` any text typed by the user after the connection is made may be replayed in their shell.  This happened to me when the connection timed out.  This could be dangerous (e.g. if someone types `rm important-file`).
2. The QUIC demo applications no longer execute successfully for me.  This seems to be due to a service change on `www.example.org` on 23 Oct (see the attached file for more details).

**Comment by hlandau** [comment]
>The QUIC demo applications no longer execute successfully for me. This seems to be due to a service change on www.example.org on 23 Oct (see the attached file for more details).

Heh &mdash; I wasn't even aware www.example.org worked, I used it because it's a standard convention to use that domain as a placeholder in documentation, not as a suggestion to actually try it. Try `www.cloudflare.com:443`.

**Comment by jamuir** [comment]
No luck with `www.cloudflare.com`:

```
demos/guide$ LD_LIBRARY_PATH=../.. ./quic-client-block 
Failed to connect to the server
40376BBD9E7F0000:error:0A0000CF:SSL routines:quic_do_handshake:protocol is shutdown:ssl/quic/quic_impl.c:1694:
```

Note that the comments in the demo Makefile suggest users can execute the applications:

```
$ head -n 7 demos/guide/Makefile 
#
# To run the demos when linked with a shared library (default) ensure that
# libcrypto and libssl are on the library path. For example to run the
# tls-client-block demo:
#
#    LD_LIBRARY_PATH=../.. ./tls-client-block

```

The TLS demos work fine against `www.example.com`.

I think there is value in having a real server to talk to when you are just learning about QUIC.  We discussed this some at today's stand-up and there were a few suggestions.  I have opened #261 for this.

**Comment by jamuir** [comment]
Matt C noted that there is a ticket open to update the demos to accept host, port, alpn as command-line args ( #252 ).  If those args are mandatory, then that would avoid the problem of users (like me :-) )  thinking that they should be able to talk to `www.example.com`.

**Comment by mattcaswell** [comment]
Your notes in the attached markdown are perfect and very interesting. It clearly shows where we need to add some clearer explanations and examples.

**Comment by hlandau** [comment]
> No luck with `www.cloudflare.com`:
> 
> ```
> demos/guide$ LD_LIBRARY_PATH=../.. ./quic-client-block 
> Failed to connect to the server
> 40376BBD9E7F0000:error:0A0000CF:SSL routines:quic_do_handshake:protocol is shutdown:ssl/quic/quic_impl.c:1694:
> ```

It doesn't really make sense to use quic-client-block for Cloudflare as it will only talk HTTP/3 (and ALPN will not negotiate correctly). Try the HTTP/3 demo?

Of course this makes it clear we need to make this clearer.

**Comment by jamuir** [comment]
I have gone through the files under `doc/designs/ddd` and made notes in the attached markdown file.
[jm-journal-part2.md](https://github.com/openssl/project/files/13191107/jm-journal-part2.md)

I have opened a PR with minor changes to the makefile and the demo sources:
https://github.com/openssl/openssl/pull/22542

I made some suggestions about updating `README.md` and will open a separate PR for those.

One of TLS demo applications, `ddd-01`, did not work all.  I fixed it in the PR.

Other TLS demos mostly work, but communicating with `www.google.com:443` exposes some problems.  E.g. `ddd-06` aborts with a failed assert.  I did not figure that one out.

The ddd QUIC applications do not work, but I guess that is expected since they need a server to talk to.

**Comment by nhorman** [comment]
[nh-journal-part1.md](https://github.com/openssl/project/files/13208708/nh-journal-part1.md)

Notes from my first day attempting to go through the quic demos/docs

Summary:
* Generally the s_client example from README-QUIC.md worked well.  might be nice to specify h3 as the alpn in the doc, instead of specifying myalpn, then indicating that it should be changed to h3, but I also could have read ahead before I tried
* working with the ddd demos was a struggle
* found one bug in the ddd-01-conn-blocking demo, in which the host:port string was constructed with a newline that messed up BIO_lookup_ex - fixed in https://github.com/openssl/openssl/pull/22553
* After a bit more alpn modifications, I was able to get a successful connection to www.example.com, but no useful data was exchanged
* Chat discussion revealed that the ddd demos weren't actually meant to be run, but provided as illustrative example as to different api usage patterns could be used to establish connections to a quic server
* suggested that we augment the documentation to make this clear
* with the addition of the inflight pr (https://github.com/openssl/openssl/pull/22552) to allow command line host/port specification in the demos/guide directory, I began trying to setup quicserver to run tests with those demos.
* Quicserver encountered an odd error on my system - see https://github.com/openssl/openssl/pull/22559 for details
* with that change, quicserver starts, but quic-client-block hangs attempting to connect - will investigate further tomorrow

**Comment by jamuir** [comment]
Here are my last set of notes for the `http3` demo:

[jm-journal-part3.md](https://github.com/openssl/project/files/13209299/jm-journal-part3.md)

The http3 demo is very nice and worked well for me.  I only made some minor comments.

I have not opened any new PRs for this, although I do think a minor update to `README.md`, and possibly `Makefile`, could be done to address my comments.

**Comment by mattcaswell** [comment]
> with that change, quicserver starts, but quic-client-block hangs attempting to connect 

There is another bug in quicserver (fixed by openssl/openssl#22557). The quicserver will timeout after a while and no connection is attempted. If that happens (e.g. if you're not quick enough starting your client)  it then goes into an infinite loop and nothing can connect to it. The fix just changes things so that you can see it failed.

**Comment by nhorman** [comment]
[nh-journal-part2.md](https://github.com/openssl/project/files/13219681/nh-journal-part2.md)

My last set of notes. Summary:
* Updated my quicserver PR to stop searching getaddrinfo returned addresses when we managed to listen on a sock.  missing break statement caused a socket leak and lack of ability to connect.
* Fix up quic clients to support selecting ipv4/6 addresses to connect to.
Everything generally worked well after those fixes, docs were fairly clear, and easy to follow
